geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/data/stats/GeoMesaStats.scala

```diff
@@ -0,0 +1,167 @@
+/***********************************************************************
+* Copyright (c) 2013-2016 Commonwealth Computer Research, Inc.
+* All rights reserved. This program and the accompanying materials
+* are made available under the terms of the Apache License, Version 2.0
+* which accompanies this distribution and is available at
+* http://www.opensource.org/licenses/apache2.0.php.
+*************************************************************************/
+
+package org.locationtech.geomesa.accumulo.data.stats
+
+import com.vividsolutions.jts.geom.Envelope
+import org.geotools.data.Query
+import org.geotools.geometry.jts.ReferencedEnvelope
+import org.joda.time.{DateTimeZone, Interval}
+import org.locationtech.geomesa.accumulo.data._
+import org.locationtech.geomesa.accumulo.data.tables.RecordTable
+import org.locationtech.geomesa.utils.geotools.{CRS_EPSG_4326, wholeWorldEnvelope}
+
+/**
+ * Tracks stats for a schema - spatial/temporal bounds, number of records, etc. Persistence of
+ * stats is not part of this trait, as different implementations will likely have different method signatures.
+ */
+trait GeoMesaStats {
+
+  /**
+   * Get rough bounds for a query
+   *
+   * @param query query
+   * @return bounds
+   */
+  def estimateBounds(query: Query): ReferencedEnvelope
+
+  /**
+   * Get rough time bounds for a query
+   *
+   * @param query query
+   * @return time bounds
+   */
+  def estimateTimeBounds(query: Query): Interval
+
+  /**
+   * Rough estimate of number of features
+   *
+   * @param query query
+   * @return count of features
+   */
+  def estimateCount(query: Query): Long
+}
+
+/**
+ * Tracks stats via entries stored in metadata
+ */
+trait GeoMesaMetadataStats extends GeoMesaStats {
+
+  this: AccumuloConnectorCreator with HasGeoMesaMetadata =>
+
+  import GeoMesaStats.{allTimeBounds, decodeSpatialBounds, decodeTimeBounds, encode}
+
+  /**
+   * Get rough bounds for a query
+   * Note: we don't currently filter by the cql
+   *
+   * @param query query
+   * @return bounds
+   */
+  override def estimateBounds(query: Query): ReferencedEnvelope =
+    readSpatialBounds(query.getTypeName)
+        .map(new ReferencedEnvelope(_, CRS_EPSG_4326))
+        .getOrElse(wholeWorldEnvelope)
+
+  /**
+   * Get rough time bounds for a query
+   * Note: we don't currently filter by the cql
+   *
+   * @param query query
+   * @return time bounds
+   */
+  override def estimateTimeBounds(query: Query): Interval =
+    readTimeBounds(query.getTypeName).getOrElse(allTimeBounds)
+
+  /**
+   * Rough estimate of number of features
+   * Note: we don't currently filter by the cql
+   *
+   * @param query query
+   * @return count of features
+   */
+  override def estimateCount(query: Query): Long =
+    metadata.getTableSize(getTableName(query.getTypeName, RecordTable))
```

Comment:
ah, no actually... looks like this would throw some kind of exception in that case. I just moved this code over from AccumuloDataStore. The count stuff is getting revamped next, so I'm inclined to just leave as is for now.